### PR TITLE
[Interfaces] Improved ```get_name()``` method of hardware interfaces #api-breaking

### PR DESCRIPTION
**Pull request by Schulze18**
Hello everyone. 

Following the ideas of the open issue #691, this PR modifies the methods to get the names related to the hardware interface.

To deprecate ```get_full_name()```, I just made it return ```get_name()```. If this is not the proper way to deprecate it, just let me know.

As I was not sure if the classes [ResourceStorage](https://github.com/ros-controls/ros2_control/blob/5363e3a248e5fc6379e44cd50d6fe59cc05514a3/hardware_interface/src/resource_manager.cpp#L66) and [ResourceManager](https://github.com/ros-controls/ros2_control/blob/5363e3a248e5fc6379e44cd50d6fe59cc05514a3/hardware_interface/src/resource_manager.cpp#L520) inherit ReadOnlyHandle, I did not modified any occurrences of ```hardware.get_name()``` and ```component.get_name()```.

Sorry in advance if I forgot any step during the PR opening or the code modifications, it is my first PR to an open source project.

Thank you.

### CODE REVIEW

File: hardware_interface/include/hardware_interface/handle.hpp [hunk 1]
```diff
@@ -56,11 +56,14 @@ class ReadOnlyHandle
   /// Returns true if handle references a value.
   inline operator bool() const { return value_ptr_ != nullptr; }
 
-  const std::string & get_name() const { return name_; }
+  const std::string get_name() const { return name_ + "/" + interface_name_; }
 
   const std::string & get_interface_name() const { return interface_name_; }
 
-  const std::string get_full_name() const { return name_ + "/" + interface_name_; }
+  // Deprecated
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/handle.hpp:63]:
```suggestion
  [[deprecated("Replaced by get_name method, which is semantically more correct")]]
```

File: hardware_interface/include/hardware_interface/handle.hpp [hunk 2]
```diff
@@ -56,11 +56,14 @@ class ReadOnlyHandle
...
+  const std::string get_full_name() const { return get_name(); }
+
+  const std::string & get_prefix_name() const { return name_; }
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/handle.hpp:66]: Would it make sense to also rename this member to `prefix_name_` just to be clearer in the future?

```suggestion
  const std::string & get_prefix_name() const { return prefix_name_; }
```

**Review by destogl — APPROVED** [review]
It looks good. Thanks!

**Comment by destogl** [comment]
> As I was not sure if the classes [ResourceStorage](https://github.com/ros-controls/ros2_control/blob/5363e3a248e5fc6379e44cd50d6fe59cc05514a3/hardware_interface/src/resource_manager.cpp#L66) and [ResourceManager](https://github.com/ros-controls/ros2_control/blob/5363e3a248e5fc6379e44cd50d6fe59cc05514a3/hardware_interface/src/resource_manager.cpp#L520) inherit ReadOnlyHandle, I did not modified any occurrences of `hardware.get_name()` and `component.get_name()`.

This is OK and don't have to be changed. Those methods are not getting name from an interface but from a hardware component, e.g., this is a robot's name.

> Thank you.

No, thank you! 🙏 😁

**Comment by bmagyar** [comment]
holding on for this until we branch for Humble

**Comment by Schulze18** [comment]
Thank you, @destogl. I ran `pre-commit` without `-a` option. Now, I am having some `build/include_order` errors mainly related to `joint_limits_interface` and `transmission_interface`, but nothing related to the changes in this PR. I believe those errors are out of scope of this PR, right?

**Comment by destogl** [comment]
@Schulze18 we just fixed CI see we will check it and see that everything goes through.

File: controller_interface/include/controller_interface/helpers.hpp [hunk 3]
```diff
@@ -51,15 +51,15 @@ bool get_ordered_interfaces(
         // check case where:
         // (<joint> == <joint> AND <interface> == <interface>) OR <joint>/<interface> == 'full name'
         if (
-          ((name == interface.get_name()) && (interface_type == interface.get_interface_name())) ||
-          ((name + "/" + interface_type) == interface.get_full_name()))
+          ((name == interface.get_prefix_name()) && (interface_type == interface.get_interface_name())) ||
+          ((name + "/" + interface_type) == interface.get_name()))
```
Comment (destogl) [inline controller_interface/include/controller_interface/helpers.hpp:55]:
```suggestion
          ((name == interface.get_prefix_name()) &&
           (interface_type == interface.get_interface_name())) ||
          ((name + "/" + interface_type) == interface.get_name()))
```

File: hardware_interface/include/hardware_interface/handle.hpp [hunk 4]
```diff
@@ -56,11 +56,14 @@ class ReadOnlyHandle
...
   inline operator bool() const { return value_ptr_ != nullptr; }
 
-  const std::string & get_name() const { return name_; }
+  const std::string get_name() const { return name_ + "/" + interface_name_; }
 
   const std::string & get_interface_name() const { return interface_name_; }
 
-  const std::string get_full_name() const { return name_ + "/" + interface_name_; }
+  [[deprecated("Replaced by get_name method, which is semantically more correct")]]
+  const std::string get_full_name() const { return get_name(); }
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/handle.hpp:64]:
```suggestion
  [[deprecated(
    "Replaced by get_name method, which is semantically more correct")]] const std::string 
  get_full_name() const
  {
    return get_name();
  }
```

**Comment by destogl** [comment]
@Schulze18 I updated everything, so we can merge and release! Thanks for initial work and following up!

**Review by bmagyar — APPROVED** [review]

**Pull request merged**